catalog/dags/common/elasticsearch.py

```diff
@@ -0,0 +1,223 @@
+
+    response = es_conn.indices.get(
+        index=source_index,
+        # Return empty dict instead of throwing error if no index can be
+        # found. We raise our own error instead.
+        ignore_unavailable=True,
+    )
+
+    if len(response) != 1:
+        raise ValueError(f"Index {source_index} could not be uniquely identified.")
+
+    # The response has the form:
+    #   { index_name: index_configuration }
+    # However, since `source_index` can be an alias rather than the index name,
+    # we do not necessarily know the index_name so we cannot access the configuration
+    # directly by key. We instead get the first value from the dict, knowing that we
+    # have already ensured in a previous check that there is exactly one value in the
+    # response.
+    config = next(iter(response.values()))
+    return config
+
+
+def remove_excluded_index_settings(index_config):
+    """
+    Remove fields from the given index configuration that should not be included when
+    using it to create a new index.
+    """
+    # Remove fields from the current_index_config that should not be copied
+    # over into the new index (such as uuid)
+    for setting in EXCLUDED_INDEX_SETTINGS:
+        index_config.get("settings", {}).get("index", {}).pop(setting)
+
+    # Aliases should also not by applied automatically
+    index_config.pop("aliases")
+
+    return index_config
+
+
+@task
+def create_index(index_config, es_host: str):
+    es_conn = ElasticsearchPythonHook(hosts=[es_host]).get_conn
+
+    new_index = es_conn.indices.create(**index_config)
+
+    return new_index
+
+
+@task_group(group_id="trigger_and_wait_for_reindex")
+def trigger_and_wait_for_reindex(
+    destination_index: str,
+    source_index: str,
+    query: dict,
+    timeout: timedelta,
+    requests_per_second: int,
+    es_host: str,
+    max_docs: int | None = None,
+    slices: Union[int, Literal["auto"]] = "auto",
+):
+    @task
+    def trigger_reindex(
+        es_host: str,
+        destination_index: str,
+        source_index: str,
+        query: dict,
+        requests_per_second: int,
+        max_docs: int | None,
+        slices: Union[int, Literal["auto"]],
+    ):
+        es_conn = ElasticsearchPythonHook(hosts=[es_host]).get_conn
+        source = {"index": source_index}
+        # An empty query is not accepted; only pass it
+        # if a query was actually supplied
+        if query:
+            source["query"] = query
+
+        response = es_conn.reindex(
+            source=source,
+            dest={"index": destination_index},
+            max_docs=max_docs,
+            # Parallelize indexing when not None
+            slices=slices,
+            # Do not hold the slot while awaiting completion
+            wait_for_completion=False,
+            # Immediately refresh the index after completion to make
+            # the data available for search
+            refresh=True,
```

Comment:
Should this be a configurable? I imagine for standard reindices this would hold true, but for the proportional-by-source, it might mean several dozen refreshes (one after each source is done being indexed). If we went that route, we might need a separate `refresh` step at the very end once indexing is all done.

Reply:
That's a good optimization, for sure! I updated it to make this configurable, set it to `False` for this DAG in particular, and added a final `refresh` task after the reindexing is complete.